src/shogun/base/SGObject.cpp

```diff
@@ -657,27 +677,65 @@ void CSGObject::get_parameter_incremental_hash(uint32_t& hash, uint32_t& carry,
 	}
 }
 
-void CSGObject::build_gradient_parameter_dictionary(CMap<TParameter*, CSGObject*>* dict)
+std::vector<std::pair<std::string, AnyParameter>> CSGObject::get_parameters_by_property(ParameterProperties property)
 {
-	for (index_t i=0; i<m_gradient_parameters->get_num_parameters(); i++)
+    std::vector<std::pair<std::string, AnyParameter>> result;
+
+	for (auto it = self->map.begin(); it != self->map.end(); ++it )
 	{
-		TParameter* p=m_gradient_parameters->get_parameter(i);
-		dict->add(p, this);
+		if (it->second.get_properties().has_property(property))
+		{
+			if(!(it->second.get_value().empty()))
+            {
+                result.push_back(std::make_pair(it->first.name(), it->second));
+			}
+		}
 	}
 
-	for (index_t i=0; i<m_model_selection_parameters->get_num_parameters(); i++)
-	{
-		TParameter* p=m_model_selection_parameters->get_parameter(i);
-		CSGObject* child=*(CSGObject**)(p->m_parameter);
+	return result;
+}
 
-		if ((p->m_datatype.m_ptype == PT_SGOBJECT) &&
-				(p->m_datatype.m_ctype == CT_SCALAR) &&	child)
+AnyParameter* CSGObject::get_parameter_by_name(std::string name)
+{
+	for (auto it = self->map.begin(); it != self->map.end(); ++it )
+	{
+		if (it->first.name().compare(name))
 		{
-			child->build_gradient_parameter_dictionary(dict);
+			return &it->second;
 		}
 	}
 }
 
+void CSGObject::build_gradient_parameter_dictionary(CMap<AnyParameter*, CSGObject*>* dict)
+{
+    std::vector<std::pair<std::string, AnyParameter>> results = get_parameters_by_property(ParameterProperties::GRADIENT);
+    std::pair<std::string, AnyParameter> *parameter;
+
+    auto found_it_lambda = [&](auto value) {
+		auto temp_val = get(parameter->first);
+		temp_val->build_gradient_parameter_dictionary(dict);
+    };
+
+    for (auto i:results)
+    {
+        dict->add(&(i.second), this);
+    }
+
+    results = get_parameters_by_property(ParameterProperties::HYPER);
```

Comment:
this could be `results = self->filter(ParameterProperties::HYPER)` and drop the `get_parameters_by_property`?